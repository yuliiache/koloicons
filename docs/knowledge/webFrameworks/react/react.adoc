== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







