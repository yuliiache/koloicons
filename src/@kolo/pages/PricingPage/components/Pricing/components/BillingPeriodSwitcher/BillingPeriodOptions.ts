import {BillingInterval} from '@kolo/constants/constants';
import {Option} from '@kolo/uiKit/RadioGroup/RadioButton/RadioButton';

export const options: Option[] = [
  {
    id: 'billed_monthly',
    name: 'billingInterval',
    value: BillingInterval.MONTH,
    label: 'Billed Monthly',
  },
  {
    id: 'billed_yearly',
    name: 'billingInterval',
    value: BillingInterval.YEAR,
    label: 'Billed Yearly',
  },
];
