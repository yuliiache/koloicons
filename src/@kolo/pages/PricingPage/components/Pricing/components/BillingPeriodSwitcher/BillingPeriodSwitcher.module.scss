@import "src/@kolo/design";

.sale {
  display: flex;
  align-items: center;
  font-size: 0.6rem;
  margin-left: 0.4rem;
  color: #525866;
}

.pricingOptions {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: center;
  margin: 0 auto 2.4rem;
  font-size: 0.6rem;
}

.saleIcon {
  margin-right: 0.4rem;
}

@media screen and (max-width: $mobileWidth) {
  .sale {
    display: inline-block;
    width: 8.45rem;
    line-height: 1;
    text-align: right;
    background: none;
    margin-top: 0.1rem;
    padding: 0;
  }

  .pricingOptions {
    flex-direction: column;
    margin-bottom: 1.2rem;
  }

  .saleIcon {
    display: none;
  }
}
