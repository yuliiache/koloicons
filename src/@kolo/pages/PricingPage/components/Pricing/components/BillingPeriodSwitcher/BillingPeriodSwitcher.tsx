import {BillingInterval} from '@kolo/constants/constants';
import Icon from '@kolo/uiKit/Icons';
import RadioGroup from '@kolo/uiKit/RadioGroup/RadioGroup';
import React from 'react';

import useBillingPeriod from '../../hooks/useBillingPeriod';
import {options} from './BillingPeriodOptions';
import styles from './BillingPeriodSwitcher.module.scss';
const {sale, saleIcon, pricingOptions} = styles;

interface Props {
  onChange: (value: BillingInterval) => void;
}
const BillingPeriodSwitcher: React.FC<Props> = ({onChange}) => {
  const {selectedBillingInterval, setSelectedBillingInterval} = useBillingPeriod();

  const handleOnChange = (value: string) => {
    setSelectedBillingInterval(value as BillingInterval);
    onChange(value as BillingInterval);
  };

  return (
    <div className={pricingOptions}>
      <RadioGroup
        options={options}
        currentValue={selectedBillingInterval}
        onChange={handleOnChange}
      />
      <span className={sale}>
        <div className={saleIcon}>
          <Icon.Arrow />
        </div>
        Save 75%
      </span>
    </div>
  );
};

export default BillingPeriodSwitcher;
