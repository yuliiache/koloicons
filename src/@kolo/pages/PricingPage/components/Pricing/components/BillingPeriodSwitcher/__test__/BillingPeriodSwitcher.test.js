import React from 'react';
import {options} from '@kolo/pages/PricingPage/components/Pricing/components/BillingPeriodSwitcher/BillingPeriodOptions.ts';
import {fireEvent, render, screen} from '@testing-library/react';

import '@testing-library/jest-dom';

import BillingPeriodSwitcher from '../BillingPeriodSwitcher';

describe('BillingPeriodSwitcher', () => {
  const mockOnChange = jest.fn();

  beforeEach(() => {
    render(<BillingPeriodSwitcher onChange={mockOnChange} />);
  });

  it('renders correctly', () => {
    options.forEach((option) => {
      expect(screen.getByLabelText(option.label)).toBeInTheDocument();
    });
    expect(screen.getByText(/save 75%/i)).toBeInTheDocument();
  });

  it('calls onChange prop with correct value when selection changes', () => {
    const newSelection = options[0].value;
    fireEvent.click(screen.getByLabelText(options[0].label));

    expect(mockOnChange).toHaveBeenCalledWith(newSelection);
  });

  it('renderes snapshot successfully', () => {
    const {asFragment} = render(<BillingPeriodSwitcher onChange={mockOnChange} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
