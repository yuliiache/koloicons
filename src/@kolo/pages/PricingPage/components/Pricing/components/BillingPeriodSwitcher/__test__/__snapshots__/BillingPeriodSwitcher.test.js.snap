// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BillingPeriodSwitcher renderes snapshot successfully 1`] = `
<DocumentFragment>
  <div
    class="pricingOptions"
  >
    <div
      class="radioGroup"
    >
      <div
        class="radioButton"
        tabindex="0"
      >
        <input
          class="radioButtonInput"
          data-testid="radioGroupComponent"
          id="billed_monthly"
          name="billingInterval"
          type="radio"
          value="month"
        />
        <label
          class="radioButtonLabel"
          for="billed_monthly"
        >
          Billed Monthly
        </label>
      </div>
      <div
        class="radioButton radioButtonActive"
        tabindex="0"
      >
        <input
          checked=""
          class="radioButtonInput"
          data-testid="radioGroupComponent"
          id="billed_yearly"
          name="billingInterval"
          type="radio"
          value="year"
        />
        <label
          class="radioButtonLabel"
          for="billed_yearly"
        >
          Billed Yearly
        </label>
      </div>
    </div>
    <span
      class="sale"
    >
      <div
        class="saleIcon"
      >
        <svg
          data-testid="icon-arrow"
          height="9"
          viewBox="0 0 27 9"
          width="27"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.16981132,0 L0,4.5 L9.16981132,9 L9.16981132,5 L10.6981132,5 L10.6981132,4 L9.16981132,4 L9.16981132,0 Z M12.7358491,4 L12.7358491,5 L14.7735849,5 L14.7735849,4 L12.7358491,4 Z M16.8113208,4 L16.8113208,5 L18.8490566,5 L18.8490566,4 L16.8113208,4 Z M20.8867925,4 L20.8867925,5 L22.9245283,5 L22.9245283,4 L20.8867925,4 Z M24.9622642,4 L24.9622642,5 L27,5 L27,4 L24.9622642,4 Z"
            fill="#919EAB"
          />
        </svg>
      </div>
      Save 75%
    </span>
  </div>
</DocumentFragment>
`;
